src/main/client_context.cpp

```diff
@@ -73,8 +74,39 @@ struct ActiveQueryContext {
 	BaseQueryResult *open_result = nullptr;
 };
 
+#ifdef DEBUG
+struct DebugClientContextState : public ClientContextState {
+	~DebugClientContextState() override {
+		D_ASSERT(active_transaction == 0);
+		D_ASSERT(active_query == 0);
+	}
+
+	idx_t active_transaction = 0;
+	idx_t active_query = 0;
+
+	void QueryBegin(ClientContext &context) override {
+		active_query++;
+	}
+	void QueryEnd(ClientContext &context) override {
+		active_query--;
+	}
+	void TransactionBegin(MetaTransaction &transaction, ClientContext &context) override {
+		active_transaction++;
+	}
+	void TransactionCommit(MetaTransaction &transaction, ClientContext &context) override {
+		active_transaction--;
```

Comment:
Similar to the QueryEnd comment, can we assert here that `active_transaction == 1` (also in TransactionRollback).